crates/integrations/datafusion/src/catalog.rs

```diff
@@ -0,0 +1,94 @@
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use std::{any::Any, collections::HashMap, sync::Arc};
+
+use datafusion::catalog::{schema::SchemaProvider, CatalogProvider};
+use futures::future::try_join_all;
+use iceberg::{Catalog, NamespaceIdent, Result};
+
+use crate::schema::IcebergSchemaProvider;
+
+/// Provides an interface to manage and access multiple schemas
+/// within an Iceberg [`Catalog`].
+///
+/// Acts as a centralized catalog provider that aggregates
+/// multiple [`SchemaProvider`], each associated with distinct namespaces.
+pub struct IcebergCatalogProvider {
+    /// A concurrent `HashMap` where keys are namespace names
+    /// and values are dynamic references to objects implementing the
+    /// [`SchemaProvider`] trait.
+    schemas: HashMap<String, Arc<dyn SchemaProvider>>,
+}
+
+impl IcebergCatalogProvider {
+    /// Asynchronously tries to construct a new [`IcebergCatalogProvider`]
+    /// using the given client to fetch and initialize schema providers for
+    /// each namespace in the Iceberg [`Catalog`].
+    ///
+    /// This method retrieves the list of namespace names
+    /// attempts to create a schema provider for each namespace, and
+    /// collects these providers into a concurrent `HashMap`.
+    pub async fn try_new(client: Arc<dyn Catalog>) -> Result<Self> {
+        let schema_names: Vec<_> = client
+            .list_namespaces(None)
+            .await?
+            .iter()
+            .flat_map(|ns| ns.as_ref().clone())
+            .collect();
+
+        let providers = try_join_all(
+            schema_names
+                .iter()
+                .map(|name| {
+                    IcebergSchemaProvider::try_new(
+                        client.clone(),
+                        NamespaceIdent::new(name.clone()),
+                    )
+                })
+                .collect::<Vec<_>>(),
+        )
+        .await?;
+
+        let schemas: Vec<_> = schema_names
+            .into_iter()
+            .zip(providers.into_iter())
+            .map(|(name, provider)| {
+                let provider = Arc::new(provider) as Arc<dyn SchemaProvider>;
+                (name, provider)
+            })
+            .collect();
+
+        Ok(IcebergCatalogProvider {
+            schemas: schemas.into_iter().collect(),
+        })
+    }
+}
+
+impl CatalogProvider for IcebergCatalogProvider {
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+
+    fn schema_names(&self) -> Vec<String> {
+        self.schemas.keys().cloned().collect()
```

Comment:
I'm thinking that this maybe incorrect since others processes may create new namespaces after creating the `try_new` method? We should the result of `list_namespaces` each time. For performance issue, we may create sth like [`CachingCatalog` in java](https://github.com/apache/iceberg/blob/c07f2aabc0a1d02f068ecf1514d2479c0fbdd3b0/core/src/main/java/org/apache/iceberg/CachingCatalog.java#L46) to implement `Catalog` trait, what do you think?

Reply:
yes, you're right this is a simple but naive impl. based on the [docs ](https://docs.rs/datafusion/latest/datafusion/catalog/trait.CatalogProvider.html#implementing-remote-catalogs) and the ref from delta-rs. 

I think having a caching wrapper would be a better solution. 

However, since we have to provide a vec of all schema_names we cannot check if a single schema is in the cache and if not fetch again. So I'm guessing we can only implement it with a time based eviction policy i.e. cache all schemas for 1min in order to avoid multiple network calls in a short amount of time (same concept as debouncing)?

Reply:
Yes, when adding a cache, we need to a design for the trade of between performance and consistency. But I think we could leave it to actualy cache implentation? For this pr, we should not cache them in the providers, e.g. `CatalogProvider`, `SchemaProvider`, etc, but calling `Catalog` api directly.

Reply:
sure, we can do it like that for this PR and impl. the more sophisticated solution later.

Reply:
@liurenjie1024 
...I took another look; and I'm not sure we can call the Catalog API directly each time the funtion on the e.g. `trait CatalogProvider` is invoked. The trait is required to be synchronous; however our catalog api calls are async.

So perhaps; we leave it as is for this PR and work on the cache instead of wasting time and effort on another intermediate solution? But perhaps there is an obvious solution to this I'm missing?

Reply:
Yes, you are right. As of now, we can only provide a snapshot version. It's possible to block on an async call, but it requires runtime api. cc @Xuanwo @Fokko @viirya What dou you think? My suggestion is to leave it as it now, and fix it when we have runtime api landed.

Reply:
> when we have runtime api landed

is there any issue or discussion already related to that? Or would you mind to quickly outline whats the plan here?

Reply:
> It's possible to block on an async call, but it requires runtime api.

It does possible but we should avoid using `rt.block_on()` in our lib. 

> My suggestion is to leave it as it now, and fix it when we have runtime api landed.

I'm second with this suggestion. We can move on and figure how to improve this part in the future.

Reply:
We should be hesitant with caching, especially avoiding upfront optimizations. For Iceberg in general, consistency is king.

> For this pr, we should not cache them in the providers, e.g. CatalogProvider, SchemaProvider, etc, but calling Catalog api directly.

This makes sense, but I see the issue with blocking on async calls. At first, I would take the price of waiting for the blocking calls. Even it is still a remote call, the ones to the REST catalog should be lightning-fast (that's where the caching happens).